up-transport-vsomeip/src/transport_inner/transport_inner_handle.rs

```diff
@@ -0,0 +1,880 @@
+                    registration_type.client_id()
+                ),
+            ));
+        };
+
+        let (tx, rx) = oneshot::channel();
+        trace!("attempting to block_on");
+
+        // Using block_in_place to perform async operation in sync context
+        let send_to_engine_res = task::block_in_place(|| {
+            CB_RUNTIME.block_on(Self::send_to_engine_with_status(
+                &self.engine.transport_command_sender,
+                TransportCommand::UnregisterListener(
+                    src,
+                    sink,
+                    registration_type.clone(),
+                    app_name,
+                    tx,
+                ),
+            ))
+        });
+
+        trace!("after attempting to block_on");
+        if let Err(err) = send_to_engine_res {
+            panic!("engine has stopped! unable to proceed! with err: {err:?}");
+        }
+        let await_engine_res =
+            task::block_in_place(|| CB_RUNTIME.block_on(Self::await_engine("unregister", rx)));
+        if let Err(warn) = await_engine_res {
+            warn!("{warn}");
+        }
+
+        let comp_listener = ComparableListener::new(listener);
+        let listener_config = (source_filter.clone(), sink_filter.cloned(), comp_listener);
+        let client_usage_res = self
+            .get_storage()
+            .get_message_handler_registry()
+            .release_message_handler(listener_config);
+
+        // TODO: We should probably also remove entries from:
+        //  * rpc_correlation -> send an error
+        //  * vsomep_offered_requested -> unoffer / unrequest
+
+        let Ok(client_usage) = client_usage_res else {
+            warn!("{}", client_usage_res.err().unwrap());
+            return Ok(());
+        };
+
+        match client_usage {
+            ClientUsage::ClientIdInUse => {}
+            ClientUsage::ClientIdNotInUse(client_id) => {
+                task::block_in_place(|| CB_RUNTIME.block_on(self.shutdown_vsomeip_app(client_id)))?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) async fn send(&self, message: UMessage) -> Result<(), UStatus> {
+        let attributes = message.attributes.as_ref().ok_or(UStatus::fail_with_code(
+            UCode::INVALID_ARGUMENT,
+            "Missing uAttributes",
+        ))?;
+
+        // Validate UAttributes before conversion.
+        UAttributesValidators::get_validator_for_attributes(attributes)
+            .validate(attributes)
+            .map_err(|e| {
+                UStatus::fail_with_code(UCode::INTERNAL, format!("Invalid uAttributes, err: {e:?}"))
+            })?;
+
+        trace!("Sending message with attributes: {:?}", attributes);
+
+        let Some(source_filter) = message.attributes.source.as_ref() else {
+            return Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "UMessage provided with no source",
+            ));
+        };
```

Comment:
Fair. Was trying to extract the source_filter in a safe way without YOLO'ing an unwrap even if it had been validated to exist.

Do you know of a better way?

Reply:
I get your point. Using something like `unwrap` or `expect` should work for now but when the validators are changed, it might not be obvious that this code would need to be changed as well ...

Reply:
Going to leave for now based on our discussion.

Reply:
I found this interesting read about the problem at hand: https://blog.burntsushi.net/unwrap/

Based on this, I would say it should be ok to simply use `unwrap()` here, but I'll leave it to your judgement ...